Cython/Compiler/Nodes.py

```diff
@@ -3340,8 +3340,15 @@ def generate_function_body(self, code):
         for arg in self.args:
             if arg.hdr_type and not (arg.type.is_memoryviewslice or
                                      arg.type.is_struct or
-                                     arg.type.is_complex):
+                                     arg.type.is_complex or
+                                     arg.type.is_cpp_class):
                 args.append(arg.type.cast_code(arg.entry.cname))
+            elif arg.hdr_type and arg.type.is_cpp_class:
```

Comment:
This condition is simpler than the long one above, so I'd move the case first. First special cases, then the general handler.